### ISSUE TITLE
Set editor header levels to n+1 to avoid multiple H1s on the page

### ISSUE THREAD

**Issue by gabestein** [pubpub/pubpub#2303]
### What's your enhancement request?

- In the PubPub editor, "Header 1" should generate `<h2>` tags, "Header 2" `<h3>` tags, and so on, so that the only h1 on the page is the Pub title.
- CSS should be updated to reflect the bumped styles
- Imports should automatically enforce this as well.
- Ideally, this would work retroactively for old Pubs if possible. But not a requirement.
- Ideally, exports to pandoc formats would respect the UX designation rather than the html element (e.g., Header 1 -> Header 1), but not a requirement.

### What would this enhancement help you accomplish?
Alignment with accessibility best practices. Our assumptions about document outline have been mis-informed. See: https://www.a11yproject.com/posts/how-to-accessible-heading-structure/#one-h1

Note: this is not an SEO-related request. I think it's important to make that point, because the two are often linked together when, in fact, the [SEO argument](https://www.stanventures.com/blog/multiple-h1-tags/) is outdated. So it's much stronger coming from a pure accessibility standpoint without conflating the two.

---

### Who requested this? (If not you)
Several folks over the years, most recently @kyledenlinger in #2301

**Comment by gabestein** [comment]
Look & feel stays the same across h1 -> h6

-heading 1 element in prosemirror that maps to heading 2 in HTML
- header 6 -> role="heading" with same styles

**Comment by idreyn** [comment]
I read the article linked above, and several that fan out from its own links:

- [How-to: Accessible heading structure](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/#one-h1)
- [Computer says NO to HTML5 document outline](https://html5doctor.com/computer-says-no-to-html5-document-outline/)
- [A decade of heading backwards](https://codepen.io/stevef/post/a-decade-of-heading-backwards)
- [There Is No Document Outline Algorithm](https://adrianroselli.com/2016/08/there-is-no-document-outline-algorithm.html) (I think this one is the most lucid and informative)
- [Why You Should Choose HTML5 `article` Over `section`](https://www.smashingmagazine.com/2020/01/html5-article-section/)

Having done so, I want to explain how I believe there are two questions here that have become muddled together, and there is actually one more layer of the misinformation onion to peel back. The questions are:

> **(1)** Is it possible to use multiple `<h1>` tags on the page to produce a document that works for screenreaders?

and then:

> **(2)** Is it possible to use _only_ `<h1>` tags, in combination with the `<section>` or `<hgroup>` tags, to produce a document that works for screenreaders?

In summary, we are trying to answer **(1)**, but the real issue being debated in these articles is **(2)**.

## Some background on **(2)**

The reason that **(2)** is being debated is because there is part of the WHATWG HTML5 specification called the "document outline algorithm" that promised that, essentially, "`<h1>` is all you need". It's easiest to look at an example:

```html
 <h1>Apples</h1>
 <p>Apples are fruit.</p>
 <section>
  <h1>Taste</h1>
  <p>They taste lovely.</p>
  <section>
   <h1>Sweet</h1>
   <p>Red apples are sweeter than green ones.</p>
  </section>
 </section>
 <section>
  <h1>Color</h1>
  <p>Apples come in various colors.</p>
 </section>
```

In this document, even though all the headings are `<h1>`, the "document outline algorithm" would use the nesting of the `<section>` tags to infer that the semantically correct nesting of the document is this:

```md
- Apples
  -  Taste
    - Sweet
  - Color
```

You can see why this would be really useful. It would let developers build reusable components that used `<h1>` across the board without worrying about their context in the application. However, this specification:

- Was never adopted by browsers
- Is also not honored by screenreaders (with a few exceptions)
- Was dropped from the WHATWG spec in 2022.

In summary, these posts take pains to make it clear to developers that (despite what they may have read elsewhere) the "document outline algorithm" does not exist, never existed in practice, and cannot be relied on to produce a nested heading structure from `<h1>`s alone. It will, as the first article claims, produce a "flat and meaningless structure" for the document.

These posts, with their historical context-gathering, citations of the WHATWG spec, and actual screenreader testing, are incredibly valuable. But I still believe this is irrelevant to PubPub.

## Why I don't believe this matters

This cannot matter to PubPub because our editor does not provide a way to create `<section>` tags. If our users want meaningful document outlines, it's up to them to use headings correctly — for instance, by never nesting a Heading 3 directly beneath a Heading 1. In the paradigm of word processing, that's about the best we can do.

Here's the context of my earlier pull-quote from the first article, "How-to: Accessible heading structure", provided by The A11y Project:

> Having multiple h1s on a web page is bad practice, for accessibility and for SEO. **It creates a flat and meaningless structure.** But, it's not a WCAG 2 AA violation. The long answer involves a short history lesson.

> Using multiple h1s was proposed with the introduction of the HTML Document Outline in HTML5. Each section in the outline could contain an H1. But here's the snag: The Document outline was never adapted by browsers or assistive technology. Therefore the support for multiple h1s was dropped in HTML5.1

_This_ is the conflation of **(1)** and **(2)**. If you swap out _"having multiple h1s"_ for _"relying solely on h1s"_, the claim is correct. As is, it implies that the mere presence of multiple `<h1>` tags is enough to confuse screenreaders; though it doesn't explain precisely how, you could imagine some failure modes:

- The screenreader assumes there is exactly one `<h1>` and fails to pick up _any_ structure after the second one
- The presence of a second `<h1>` implicitly collapses all headings down to the same level for the screenreader.

In my testing with VoiceOver, neither of these things are true. Here's what I did:

- Created [a Pub](https://demo.duqduq.org/pub/h8yeqed1/release/1) with several `<h1>` elements in the body.
- Opened VoiceOver and used Ctrl+Alt+U to trigger the _Rotor_ menu, which provides a list of all headings on the page.
- Verified that the Rotor reported a correct list of all headings and their levels.

Now, VO has only a small share of the desktop screenreader market — in practice, most users prefer a much more full-featured and Windows-only program called JAWS, which I do not have ready access to. However, I am confident in saying that JAWS is _at least_ as capable as VO in this domain. There is the possibility that there are advanced, JAWS-only navigation features that fail in the presence of multiple `<h1>` tags, but nobody points to one. And the only self-identifying screenreader user who I found dipping their toe into this debate [thinks it's fine](https://stackoverflow.com/questions/35314891/do-multiple-h1s-cause-screen-readers-problems#comment102637865_58107524) to use as many `<h1>` elements as you want.

## My conclusions

- Basic heading-related screenreader functionality works fine with multiple `<h1>` elements in a Pub.
- If there is more advanced screenreader functionality out there that relies on the presence of a single `<h1>`, I don't know about it, and I don't believe any of the authors of the linked posts do either.
- The A11y Project's advice to _only use one `<h1>`_ is incorrectly derived from some great research on why it's not okay to _only use `<h1>`_. Although it's a _safe_ recommendation, I think it's a poor one, as it muddles the issue and detracts from frankly more pressing a11y issues on the web.

Since users ask about this frequently, I'm not opposed to proceeding with this work. I'm also open to finding out I'm completely wrong about this — because of the way JAWS works, or the documented preference of screenreader users, or something else. But if we wanted to push back on this change as a clear accessibility improvement, I think we have the grounds to do so. And, in any case, I wanted to document my findings on the issue.

**Comment by qweliant** [comment]
We ran through solutions like getting rid of h1s in our base prosemirror schema or allowing communities to set the usage of header tags but when people paste into the editor there becomes an issue with the amount of complexity over the editor we will need to create over the h1 tag

**Comment by gabestein** [comment]
Closing as not planned. See Ian's conclusion above.

In summary:

Most of the debates over H1s and screen reader compatibility involve whether screen readers can correctly interpret nested headers on a page that contains only H1s and uses sections to signal header level (the answer is no). This is moot for PubPub, which doesn't support it. You can use multiple H1s on a page and produce viable document outlines as long as you intend them to all be at the same level, which we've confirmed both in the WCAG and in actual screen reader testing. You can also easily break outlines by incorrectly, for example, nesting an H3 below an H1. As a Word processor, we can nudge people towards best practices, but we can't reasonably prevent people from creating poorly structured documents.
